File: bin/mk-thumbs.py
#!/usr/bin/env python

# -I /home/davep/bin

import os, sys, string, re, tempfile, fnmatch
from Tkinter import *
from opts import *
import Image
import ImageTk

opts = [
    # opt-char, opt-name, if-set-val, default-val, help-string
    FlagOption('r', 'no_rotate', 1, 0, 'do not rotate'),
    
    # opt-char, opt-name, how-to-set, default-val, help-string
    ArgOption('p', 'pic_dir', None, '.' , 'pic source dir'),
    ArgOption('t', 'thumb_dir', None, '.', 'thumbnail dest dir'),
    ArgOption('v', 'view_size', None, 160, 'image view size'),
    ArgOption('s', 'thumb_size', None, 80, 'thumbname size'),
    ArgOption('f', 'filter', None, '*.jpg', 'file selection filter')
    ]

# print 'tempfile.tempdir:', tempfile.tempdir
if (tempfile.tempdir == '/var/tmp') or (tempfile.tempdir == None):
    tempfile.tempdir = '/usr/tmp'

DELETED = 'red'
DELETED_SEL = 'purple'
SELECTED = 'blue'
UNSELECTED = 'gray'

# rotated/final pics
rpics = []

def scale_pil_image(image, scale):
    (x, y) = image.size
    x = int(x * scale)
    y = int(y * scale)
    return image.resize((x, y))

class TempFile:
    def __init__(self, ext='.ppm'):
        self.pathname = tempfile.mktemp('-mkt' + ext)

    def path(self):
        return self.pathname

    def __del__(self):
        os.system('rm -f %s' % self.pathname)
        
def get_image_dimensions(path):
    cmd = 'xli -identify %s' % path
    f = os.popen(cmd)
    resp = f.read()
    print 'resp:', resp
    f.close()
    m = re.search('(\d+)x(\d+)', resp)
    if m:
        return (m.group(1), m.group(2))
    else:
        raise "Bad xli identify output."

def scale_for(path, desired_max):
    x, y = get_image_dimensions(path)
    if x>y:
        max = x
    else:
        max = y
        
    return float(desired_max) / float(max)


class ImageInfo:
    def __init__(self, path):
        self.path = path
        self.image = None
        self.tag = None
        self.photo = None
        self.deleted = None
        self.rotated = 0

    def rotate(self, deg):
        """track total rotation, normalized to 0..359"""
        self.rotated += deg
        self.rotated %= 360

def bbbutton(win, text, cmd):
    """A little helper func to make a 'standard' button"""
    b = Button(win, text=text, command=cmd)
    b.pack(side=LEFT, expand=TRUE)
    return b

class ImageList(Frame):
    def __init__(self, pic_dir, master=None):
        Frame.__init__(self, master)
        self.pic_dir = pic_dir
        self.master = master
        self.f = Frame(self)
        self.f.pack(expand=TRUE, fill=BOTH)
        self.image_refs = []
        self.scroll = Scrollbar(self.f)
        self.scroll.pack(side=LEFT, fill=Y)
        self.text = Text(self.f, setgrid=TRUE, relief=GROOVE,
                         yscrollcommand=self.scroll.set)
        self.text.pack(side=TOP, expand=TRUE, fill=BOTH)
        self.scroll['command'] = self.text.yview

        self.bbar = Frame(self.f)
        bbbutton(self.bbar, 'Quit', self.quit_fn)
        bbbutton(self.bbar, 'Done', self.done_fn)
        self.bbar.pack(side=BOTTOM, expand=FALSE, fill=X)
        
        self.text.focus()
        for k in ('Control-q',  'Control-c', 'q', 'c'):
            self.text.bind('<%s>' % k, self.exit_fn)
            
        for k in ('Control-d', 'Control-x', 'd', 'x'):
            self.text.bind('<%s>' % k, self.delete_cur)
                
        self.text.bind('<r>', self.rotate_event_90)
        self.text.bind('<l>', self.rotate_event_270)
                
        self.text.bind('<Right>', self.move_right)
        self.text.bind('<Left>', self.move_left)
        
        self.image_info = []
        self.cur_choice = None
        self.pack(expand=TRUE, fill=BOTH)

    def rotate_event_90(self, event=None):
        self.rotate(degrees=90)

    def rotate_event_270(self, event=None):
        self.rotate(degrees=270)
        
    def move_left(self, event=None):
        self.move_cur(-1)

    def move_right(self, event=None):
        self.move_cur(1)

    def done_fn(self, event=None):
        global rpics
        rpics = []
        for ii in self.image_info:
            if not ii.deleted:
                rpics.append((os.path.basename(ii.path), ii.rotated))
          
        self.quit()

    def quit_fn(self, event=None):
        sys.exit(2)

    def exit_fn(self, event=None):
        sys.exit(2)
      
    def mk_display_thumb(self, ii):
        #
        # use this if no jpeg support is compiled into PIL
        #
        # tfo = TempFile()
        # tmp_file = tfo.path()
        # cmd = 'jpegtopnm %s > %s' % (ii.path, tmp_file)
        # print cmd
        # rc = os.system(cmd)
        # if rc:
        #     raise 'cmd>%s< failed.' % cmd

        tmp_file = ii.path
        ii.pil_image = Image.open(tmp_file)
        # print 'format:', ii.pil_image.format
        # print 'mode:', ii.pil_image.mode
        # print 'info:', ii.pil_image.info
        scale = scale_for(ii.path, options.view_size)
        ii.pil_image = scale_pil_image(ii.pil_image, scale)
        ii.photo = ImageTk.PhotoImage(ii.pil_image)

    def rotate(self, image_info=None, degrees=90):
        """make a new preview thumb and display it"""
        if image_info == None:
            image_info = self.cur_choice
        if image_info == None:
            return

        image_info.pil_image = image_info.pil_image.rotate(360-degrees)
        image_info.rotate(degrees)      # keep track of total rotation
        image_info.photo = ImageTk.PhotoImage(image_info.pil_image)
        self.text.image_configure(image_info.tag, image=image_info.photo)

    def move_cur(self, num):
        if self.cur_choice:
            idx = self.image_info.index(self.cur_choice)
            idx = (idx + num) % len(self.image_info)
        else:
            idx = 0
        self.choose(self.image_info[idx])
			   
    def delete_cur(self, event=None):
        if self.cur_choice:
            self.cur_choice.deleted = not self.cur_choice.deleted
            if self.cur_choice.deleted:
                self.set_image_bg(self.cur_choice, DELETED_SEL)
            else:
                self.set_image_bg(self.cur_choice, SELECTED)

    def set_image_bg(self, image_info, bg):
        self.text.tag_configure(image_info.tag, background=bg)
      
    def choose(self, image_info):
        #
        # set bg of old current item
        #
        if self.cur_choice:
            if self.cur_choice.deleted:
                self.set_image_bg(self.cur_choice, DELETED)
            else:
                self.set_image_bg(self.cur_choice, UNSELECTED)

        self.cur_choice = image_info
      
        #
        # set bg of new current item
        #
        if self.cur_choice.deleted:
            self.set_image_bg(self.cur_choice, DELETED_SEL)
        else:
            self.set_image_bg(self.cur_choice, SELECTED)

    def tagit(self, ii):
        tag_name = ii.image
        ii.tag = tag_name
        # print 'pp>%s<, index>%s<' % (tag_name, image)
        self.text.tag_add(tag_name, '%s' % ii.image)
        # print 'self in tagit:', self
        self.text.tag_bind(tag_name, "<Double-1>",
                           lambda r, f=self.rotate, ii=ii: f(ii))
        self.text.tag_bind(tag_name, "<Button-2>",
                           lambda r, f=self.rotate, d=270, ii=ii: f(ii, 270))
        self.text.tag_bind(tag_name, "<Button-1>",
                           lambda e,f=self.choose, ii=ii: f(ii))

    def add_pics(self, pic_names):
        for pic_name in pic_names:
            pic_path = self.pic_dir + "/" + pic_name
            ii = ImageInfo(pic_path)
            self.image_info.append(ii)
            
            # print "read %s" % pic_path
            self.mk_display_thumb(ii)
            ii.image = self.text.image_create(END, image=ii.photo,
                                              padx=4, pady=4)
            # print 'new image:', image, 'image names:', self.text.image_names()
            self.tagit(ii)
            self.text.insert(END, pic_name)

    def disable_input(self):
        self.text.configure(state=DISABLED)

    def enable_input(self):
        self.text.configure(state=NORMAL)

def tkrotate(pics, pic_dir):
    """Create a tk app to display pics and allow user to select rotations."""
    master = Tk()
    ilist = ImageList(pic_dir, master)
    ilist.add_pics(pics)
    ilist.disable_input()
    master.title("thumbnailer")
    master.mainloop()
    return rpics


def list_images(dir):
    tpics = os.listdir(dir)
    # print 'dir:', dir, 'tpics:', tpics
    pics = []
    filter = options.filter
    for pic in tpics:
        # print 'ext:', pic[-4:]
        if fnmatch.fnmatch(pic, filter):
            pics.append(pic)
    del(tpics)
    pics.sort()
    return pics
    

options = Options(sys.argv, opts)

pics = list_images(options.pic_dir)
      
# get listing of thumb dir
thumbs = list_images(options.thumb_dir)

# for all in pic and not in thumb
new_pics = []
for pic in pics:
    if pic not in thumbs:
        new_pics.append(pic)
print 'pics:', pics
print 'thumbs:', thumbs
print 'new_pics:', new_pics

Image.init()

#
# Now, display them all in a tk window and allow us to select and
# rotate them
#
if not options.no_rotate:
    new_pics = tkrotate(new_pics, options.pic_dir)
    print 'new_pics:', new_pics

#
# now make real thumbnails
#

for pic, degrees in new_pics:
    sfile = options.pic_dir + "/" + pic
    dfile = options.thumb_dir + "/" + pic
    # printf("scale %s into %s\n", sfile, dfile)
    scale = scale_for(sfile, options.thumb_size)
    if degrees:
        rot_cmd = 'jpegtran -rotate %s %s | djpeg' % (degrees, sfile)
    else:
        rot_cmd = 'djpeg %s' % sfile
    print "scale %f %s\n" % (scale, sfile)
    cmd = "%s | pnmscale %s | cjpeg -quality 100 > %s" % (rot_cmd,
                                                               scale,
                                                               dfile)
    print "cmd:", cmd

    rc = os.system(cmd)
    if rc:
        raise 'cmd>%s< failed.' % cmd

sys.exit(0)



#          tfo = TempFile()
#          tmp_file = tfo.path()
#          if image_info == None:
#              image_info = self.cur_choice

#          if not image_info:
#              return

#          path = image_info.path
#          image = image_info.image

#          # rotate the original, since we want that corrected, too.
#          cmd = "jpegtran -rotate %s %s > %s" % (degrees, path, tmp_file)
#          # print 'cmd:', cmd
#          rc = os.system(cmd)
#          if rc:
#              raise 'cmd>%s< failed.' % cmd
#          cmd = "cp -f %s %s" % (tmp_file, path)
#          # print 'cmd:', cmd
#          rc = os.system(cmd)
#          if rc:
#              raise 'cmd>%s< failed.' % cmd

File: bin/mk-thumbs.rb
#!/usr/local/bin/ruby
# -I /home/davep/bin

require "tktext"
require "tempfile"
require "getoptlong"

PIC_DIR = '-p'
THUMB_DIR = '-t'
VIEW_SIZE = '-v'
THUMB_SIZE = '-s'
PRINT_OPTIONS = '-o'
NO_ROTATE = '-r'

DELETED = 'red'
DELETED_SEL = 'purple'
SELECTED = 'blue'
UNSELECTED = 'gray'

$options = {
  PIC_DIR => '/usr/tmp',
  THUMB_DIR => '/usr/tmp/thumbs',
  VIEW_SIZE => 160,
  THUMB_SIZE => 80,
  PRINT_OPTIONS => nil,
  NO_ROTATE => nil
}

# rotated/final pics
$rpics = []  
def get_tmp_file()
  f = Tempfile.new('mk-thumbs', '/usr/tmp')
  f.close()
  #p f.path
  return f
end

def image_dimensions(path)
  fail(format("no such file %s\n",path)) if !File.exist?(path)
  resp = `xli -identify #{path}`
  if resp =~ /is a (\d+)x(\d+)/
    return $1.to_i, $2.to_i
  else
    fail "Bad xli identify output."
  end
end

def scale_for(path, desired_max)
  x, y = image_dimensions(path)
  max = x > y ? x : y
  Float(desired_max) / Float(max)
end

class ImageInfo
  def initialize(path, image, tag)
    @path = path
    @image = image
    @tag = tag
    @deleted = nil
  end

  attr_reader(:path, :image, :tag)
  attr_accessor(:deleted)
end

class ImageList < TkFrame
  def initialize(pic_dir)
    @pic_dir = pic_dir
    @f = TkFrame.new(Tk.root)
    @f.pack('expand'=>'yes', 'fill'=>'both')

    @text = TkText.new(@f) {
      setgrid 'true'
      relief 'groove'
    }
    @text.pack('side'=>'left', 'expand'=>'yes', 'fill'=>'both')

    ybar = TkScrollbar.new(@f)
    ybar.pack('side'=>'left', 'expand'=>'no', 'fill'=>'y')
    @text.yscrollbar(ybar)

    @bbar = TkFrame.new(Tk.root)
    TkButton.new(@bbar) {
      text 'Quit'
      command proc{exit}
      pack('side'=>'left', 'expand'=>'yes')
    }

    d = proc{self.mydone}
    TkButton.new(@bbar) {
      text 'Done'
      command d
      pack('side'=>'left', 'expand'=>'yes')
    }
    @bbar.pack('side'=>'bottom', 'expand'=>'no', 'fill'=>'x')

    @text.focus
    for k in %w(Control-q Control-c q c)
      @text.bind(k, proc{exit})
    end
    for k in %w(Control-d Control-x d x)
      @text.bind(k, proc{self.delete_cur})
    end
    for k in %w(Control-u u)
      @text.bind(k, proc{self.delete_cur})
    end
    @text.bind('r', proc{self.rotate(nil, 90)})
    @text.bind('l', proc{self.rotate(nil, 270)})
    
    @text.bind('Right', proc{self.move_cur(1)})
    @text.bind('Left', proc{self.move_cur(-1)})

    
    @image_info = []
    @cur_choice = nil
  end

  def mydone()
    $rpics = []
    for ii in @image_info
      $rpics.push(File.basename(ii.path)) unless ii.deleted
    end
    Tk.root.destroy()
  end

  def mk_display_thumb(path)
    tmp_file0 = get_tmp_file()
    tmp_file = tmp_file0.path
    scale = scale_for(path, $options[VIEW_SIZE])
    system("djpeg #{path} | pnmscale #{scale} > #{tmp_file}")
    photo = TkPhotoImage.new('file'=>tmp_file)
    return photo
  end

  def rotate(image_info=nil, degrees=90)
    tmp_file0 = get_tmp_file()
    tmp_file = tmp_file0.path
    if image_info == nil
      image_info = @cur_choice
    end
    return if !image_info
    path = image_info.path
    image = image_info.image

    # rotate the original, since we want that corrected, too.
    cmd = "jpegtran -rotate #{degrees} #{path} > #{tmp_file}"
    #p cmd
    system(cmd)
    system("cp -f #{tmp_file} #{path}")

    # make a new preview thumb and display it
    image.image = mk_display_thumb(path)
  end

  def move_cur(num)
    if @cur_choice
      idx = @image_info.index(@cur_choice)
      idx = (idx + num) % @image_info.nitems()
    else
      idx = 0
    end
    choose(@image_info[idx])
  end
			   
  def delete_cur()
    if @cur_choice
      @cur_choice.deleted = !@cur_choice.deleted
      if @cur_choice.deleted
	@cur_choice.tag.configure('background'=>DELETED_SEL)
      else
	@cur_choice.tag.configure('background'=>SELECTED)
      end
    end
  end
      
  def choose(image_info)
    if @cur_choice
      if @cur_choice.deleted
	@cur_choice.tag.configure('background'=>DELETED)
      else
	@cur_choice.tag.configure('background'=>UNSELECTED)
      end
    end
    @cur_choice = image_info
    if @cur_choice.deleted
      @cur_choice.tag.configure('background'=>DELETED_SEL)
    else
      @cur_choice.tag.configure('background'=>SELECTED)
    end      
  end

  #
  # do this in a method so that the value
  # of ii is maintained correctly in the closure
  # for the tag bindings.
  # if it is in the add_pics method, then ii
  # is that last value in the pic_names
  # iteration
  #
  def tagit(pic_path, image)
      tag = TkTextTag.new(@text, image.path)
      ii = ImageInfo.new(pic_path, image, tag)
      @image_info.push(ii)
      tag.bind("Double-1", proc {self.rotate(ii)})
      tag.bind("Button-2", proc {self.rotate(ii, 270)})
      tag.bind("Button-1", proc {self.choose(ii)})
  end

  def add_pics(pic_names)
    for pic_name in pic_names
      pic_path = @pic_dir + "/" + pic_name
      printf("read %s\n", pic_path)
      photo = mk_display_thumb(pic_path)
      image = TkTextImage.new(@text, 'end', 
                              {'image'=>photo, 'padx'=>4, 'pady'=>4})

      tagit(pic_path, image)
    end
  end

  def disable_input()
    @text.configure('state'=>'disabled')
  end

  def enable_input()
    @text.configure('state'=>'normal')
  end

end

def tkrotate(pics, pic_dir)
  ilist = ImageList.new(pic_dir)
  ilist.add_pics(pics)
  ilist.disable_input()
  Tk.root.title("thumbnailer")
  Tk.mainloop()
  $rpics
end

begin

  parser = GetoptLong.new
  parser.set_options(
	[PIC_DIR, '--pic-dir', GetoptLong::REQUIRED_ARGUMENT],
	[THUMB_DIR, '--thumb-dir', GetoptLong::REQUIRED_ARGUMENT],
	[VIEW_SIZE, '--view-size', GetoptLong::REQUIRED_ARGUMENT],
	[THUMB_SIZE, '--thumb_size',GetoptLong::REQUIRED_ARGUMENT],
	[PRINT_OPTIONS, GetoptLong::NO_ARGUMENT])

  parser.each { |name, arg|
    $options[name] = arg
  }

  p $options if $options[PRINT_OPTIONS]

  back = Dir.pwd
  # get listing of pic dir
  Dir.chdir($options[PIC_DIR])
  pics = Dir["*.jpg"]

  Dir.chdir(back)

  # get listing of thumb dir
  Dir.chdir($options[THUMB_DIR])
  thumbs = Dir["*.jpg"]

  Dir.chdir(back)

  # for all in pic and not in thumb 
  new_pics = pics - thumbs
  #p pics
  #p thumbs
  #p new_pics

  #
  # Now, display them all in a tk window and allow us to select and
  # rotate them
  #
  new_pics = tkrotate(new_pics, $options[PIC_DIR]) unless $options[NO_ROTATE]

  #
  # now make real thumbnails
  #
  for pic in new_pics
    sfile = $options[PIC_DIR] + "/" + pic
    dfile = $options[THUMB_DIR] + "/" + pic
    #printf("scale %s into %s\n", sfile, dfile)
    scale = scale_for(sfile, $options[THUMB_SIZE])
    printf("scale %f %s\n", scale, sfile)
    cmd = "djpeg #{sfile} | pnmscale #{scale} | cjpeg -quality 100 > #{dfile}"
    #printf("%s\n", cmd)
    system(cmd)
  end

rescue
  printf($stderr, "I failed, %s.\n", Dir.pwd());
  fail
end

exit(0)
